var/spack/repos/builtin/packages/mpich/package.py

```diff
@@ -247,19 +247,19 @@ class Mpich(AutotoolsPackage, CudaPackage, ROCmPackage):
     depends_on("argobots", when="+argobots")
 
     # building from git requires regenerating autotools files
-    depends_on("automake@1.15:", when="@develop", type="build")
-    depends_on("libtool@2.4.4:", when="@develop", type="build")
-    depends_on("m4", when="@develop", type="build"),
-    depends_on("autoconf@2.67:", when="@develop", type="build")
+    depends_on("automake@1.15:", type="build", when="@develop")
+    depends_on("libtool@2.4.4:", type="build", when="@develop")
+    depends_on("m4", type="build", when="@develop"),
+    depends_on("autoconf@2.67:", type="build", when="@develop")
 
     # building with "+hwloc' also requires regenerating autotools files
-    depends_on("automake@1.15:", when="@3.3:3.3.99 +hwloc", type="build")
-    depends_on("libtool@2.4.4:", when="@3.3:3.3.99 +hwloc", type="build")
-    depends_on("m4", when="@3.3:3.3.99 +hwloc", type="build"),
-    depends_on("autoconf@2.67:", when="@3.3:3.3.99 +hwloc", type="build")
+    depends_on("automake@1.15:", type="build", when="@3.3:3.3.99 +hwloc")
+    depends_on("libtool@2.4.4:", type="build", when="@3.3:3.3.99 +hwloc")
+    depends_on("m4", type="build", when="@3.3:3.3.99 +hwloc"),
```

Comment:
```suggestion
    depends_on("m4", type="build", when="@3.3:3.3.99 +hwloc")
```

:)

Reply:
so not nondeterminism :D

Reply:
@haampie the comma was there before, though. Is that really the fix?

Reply:
Ok here's why:

```console
> spack pkg source --canonical mpich > ~/branch
> git co f6b23b4653
> spack pkg source --canonical mpich > ~/orig
> diff ~/orig ~/branch
12,13c12,13
<     (depends_on('m4', when='@develop', type='build'),)
<     (depends_on('m4', when='@3.3:3.3.99 +hwloc', type='build'),)
---
>     (depends_on('m4', type='build', when='@develop'),)
>     (depends_on('m4', type='build', when='@3.3:3.3.99 +hwloc'),)
```

There are two `depends_on("m4")` lines, both of which ends in commas due to some ancient copy-paste error.  The fact that they're tuples b/c of the commas causes them to be ignored by the directive-stripping logic during source canonicalization. So changing *these particular* directives gets you a hash change.

We should remove both commas - it won't fix the hash change this time but it'll prevent future arbitrary hash changes.

Reply:
Then submit a separate PR, merge it, see if this PR doesn't cause hash changes (at least in CI).

But again, I don't get the urgency with this pr.

Reply:
I don't have a lot of urgency with this PR.  I was originally just curious why we cared, if it was just a mechanical order change.  I like the idea of doing a separate PR to get rid of the commas. No rush.

Reply:
@adamjstewart's OCD is both Obsessive Compulsive Disorder and Other Complexity Detector.

Reply:
I think a PR like this makes sense if we enforce the order with some tool in CI, otherwise we'll start getting package updates that do it the other way (and I don't think we can require reviewers to catch all of them)

Reply:
> But again, I don't get the urgency with this pr.

No urgency from my side. In the past I've been asked to submit big changes like this before minor releases instead of after so it makes cherry-picking easier. If the people who actually need to make patch releases don't care, then I don't care.

Reply:
> I think a PR like this makes sense if we enforce the order with some tool in CI, otherwise we'll start getting package updates that do it the other way (and I don't think we can require reviewers to catch all of them)

Agreed but idk how to enforce this. Would love to enforce this for all function calls, not just `depends_on`.

Reply:
For us it shouldn't matter as we don't backport package changes. But that said, there are probably people who cherr-pick security updates of packages from $next_spack to $current_spack